File: 2021/12/15/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prehead = ListNode(0)
        prehead.next = head
        current = head
        while current != None and current.next != None:
            p = current.next
            if current.val <= p.val:
                current = p
                continue

            current.next = p.next
            pre, next = prehead, prehead.next
            while next.val < p.val:
                pre = next
                next = next.next
            pre.next = p
            p.next = next

        return prehead.next

File: 2021/12/15/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
	prehead := &ListNode{Next: head}

	current := head
	for current != nil && current.Next != nil {
		p := current.Next
		if current.Val <= p.Val {
			current = p
			continue
		}

		current.Next = p.Next
		pre, next := prehead, prehead.Next
		for next.Val < p.Val {
			pre = next
			next = next.Next
		}
		pre.Next = p
		p.Next = next
	}

	return prehead.Next
}
